### ISSUE TITLE
How to prepare for my own data?

### ISSUE THREAD

**Issue by hzwfl2** [bknyaz/graph_attention_pool#4]
Hi, I think this is a valuable work. I want to know how to create my own dataset as the format of  `mnist_75sp_train`. 
For example, I write a Pytorch `DataLoader` as follows:
   `data_loader=DataLoader(my_dataset,batch_size=batch_size,shuffle=True)`
  It returns batch_images (`shape: batch_size*H*W`) and batch_labels (`shape: batch_size*1`) in each batch. How should I do to construct my own dataset and graph for the train? Can you provide a code to help to do this? Thank you very much!

**Comment by bknyaz** [comment]
Sorry I didn't notice this issue. Are you still having this problem?
To create `mnist_75sp_train` I use the script `extract_superpixels.py`: `python extract_superpixels.py -s train -t 4`.
Let me know if you need futher help.

**Comment by sleeplessai** [comment]
@bknyaz 
Hi, I wonder that super-pixels dataset preparation is founded for node classification or graph classification ?

**Comment by bknyaz** [comment]
Hi, it is for graph classification = image classification. But in principle it can also be used for node classification, which will make it equivalent to the semantic segmentation problem.

**Comment by sleeplessai** [comment]
@bknyaz 
Thanks. I am preparing my own data with superpixel scheme and following the repo homepage instruction.
I notice these lines:
`# mean and std computed for superpixel features in the training set`
`mn = torch.tensor([0.11225057, 0.11225057, 0.11225057, 0.44206527, 0.43950436]).view(1, 1, -1)`
`sd = torch.tensor([0.2721889,  0.2721889,  0.2721889,  0.2987583,  0.30080357]).view(1, 1, -1)`

How to calculate mean and std for node features? Which step should I start to work these out?
While the `sp_intensity` and `sp_coord` in single image has been calculated,
for the whole dataset (B, C, W, H) I calculated mean and std via `np.mean/np.std(data, axis=(0, 2, 3))` thus I got:
`[0.47861066 0.49431583 0.43597046] / [0.21912035 0.22233568 0.2661024]`
Can I adopt these results for `sp_coord` and `sp_intensity` directly or should I re-calculate for each image superpixel features?

**Comment by bknyaz** [comment]
To get the `mn` and `sd` values I used this function to go over the entire dataset and collect global statistics:
https://github.com/bknyaz/graph_attention_pool/blob/master/utils.py#L188. Note that I'm using `sd[sd < 1e-2] = 1` to avoid dividing by small numbers and blowing up node features.
Then during training and testing I use this `mn` and `sd` as here: https://github.com/bknyaz/graph_attention_pool/blob/master/train_test.py#L113

So there is no need to calculate statistics for each image. The steps are:
1. compute global statistics `mn`, `sd` (this is done once before training)
2. during training/testing for each node (superpixel) with features x (`sp_coord`, `sp_intensity`) you compute `(x-mn)/sd`

But you can try computing statistics `mn`, `sd` for each image instead of using global ones. It can result in better performance in some cases. I haven't tried this in this project.
